### ISSUE TITLE
 what about support for files in archives? 

### ISSUE THREAD

**Issue by fommil** [pathikrit/better-files#121]
<deleted>

**Comment by fommil** [comment]
(btw, key to the abstract pattern is that there needs to be a clean way to close that abstract filesystem. Keeping jars/zips open is really really bad on Windows https://github.com/fommil/class-monkey )

**Comment by pathikrit** [comment]
Hi, the DNA of better-files was supposed to just add useful implicit defs to `java.util.File` and thus it is currently only serves as File I/O util using NIO. But there's no reason, it can expand its horizon and work on other filesystems - I just did not have anything immediate in my mind but PRs on this is VERY WELCOME. A lot of the other codes in the files you posted are already in e.g. `RichOutpustStream.drain` are already in better-files (see OutputStreamOps)

**Comment by fommil** [comment]
I want to move away from `File` and `{Input,Output}Stream` entirely because they are very limiting. `Path` is a far better (and more memory efficient) solution.

BTW, when using `String` as the representation of files, this causes huge memory problems in applications that track files. We monkey patch Apache VFS to workaround these limitations, but `Path` solves it because the underlying repr for a file is parent/child nodes... this way `/path/to/my/deeply/nest/project/directory` is not repeated for every single file under management.

**Comment by jvican** [comment]
@fommil This is a great idea and I can see how it would also be super useful for lots of projects I have in mind.
  
We'll be publishing this week an official artifact of better-files (/cc @pathikrit), so perhaps we can use that as the first starting point to migrate to it in Ensime and create these abstractions in a PR. What do you think?

**Comment by fommil** [comment]
The ensime code is GPL, but I'm the sole author of the files I referenced (and their tests) and I'm happy granting an Apache 2 license to either of you for them to include it here.

**Comment by fommil** [comment]
@pathikrit I **really** wish I could help out but unfortunately my time to contribute to community projects is already incredibly limited and I have a TODO list longer than I can possibly keep. I'll have to be a bit rubbish and just say: I'm here to help you to take what you want from ENSIME, but I can't commit any time to raise any PRs myself.

**Comment by jvican** [comment]
@fommil That makes perfect sense.
  
If you grant us an Apache 2 license, I could put this a Scala Center hackathon ticket that people could hack on. It wouldn't be difficult to port what you guys have in Ensime to this repository. I think the use case is really worth it. WDYT?

**Comment by fommil** [comment]
Yeah, that's fine. I don't think it will be as simple as copy/paste, the API probably needs to be designed to reflect the Path concept of NIO on virtual file systems. I'm not sure if limiting it to just File is going to be as useful as in could be.

Btw we are still missing a lot of the needed functionality, e.g. some way of streaming through the jar contents in an efficient way. My API is random access, and hence very slow for that use case (we currently stream using VFS)

**Comment by pathikrit** [comment]
@fommil: Let's not close issues without any comments

**Comment by fommil** [comment]
recreate the ticket if you want to track it. I'm closing all my scala tickets.

**Comment by pathikrit** [comment]
@fommil : Good luck with whatever you have planned! But, perhaps, don't be destructive of the past on the way out :)

**Comment by pathikrit** [comment]
Sorry, this issue is not resolved. I won't close it till it is resolved.
